python/cudf/cudf/core/index.py

```diff
@@ -178,6 +178,28 @@ def serialize(self):
     def __contains__(self, item):
         return item in self._values
 
+    @annotate("INDEX_EQUALS", color="green", domain="cudf_python")
+    def equals(self, other):
+        """
+        Determine if two Index objects contain the same elements.
+
+        Returns
+        -------
+        out: bool
+            True if “other” is an Index and it has the same elements
+            as calling index; False otherwise.
+        """
+        if not isinstance(other, Index):
+            return False
+        if isinstance(self, CategoricalIndex):
+            other = other.astype(self.dtype)
+        elif isinstance(other, CategoricalIndex):
+            self = self.astype(other.dtype)
```

Comment:
Pandas doesn't seem to follow this behavior:
```
import pandas as pd

one = pd.CategoricalIndex(['a', 'b', 'c', 'b', 'a'], categories=['a', 'b', 'c'])
two = pd.CategoricalIndex(['a', 'b', 'c', 'b', 'a'], categories=['_', 'a', 'b', 'c'])
one.equals(two)
# False
```

Reply:
Hmm, OK so we might need to have an additional specialization for when _both_ `self` and `other` are `CategoricalIndex`, in which case we don't cast.

Reply:
Sounds like we should add that specialization in a `CategoricalIndex.equals`

Reply:
Fixed, but handled it here in `Index.equals()`.

We also need to convert `self` to a `CategoricalIndex` when `other` is a `CategoricalIndex`, so handling it here seems cleaner?